src/coreclr/jit/assertionprop.cpp

```diff
@@ -3449,6 +3453,11 @@ GenTree* Compiler::optAssertionPropGlobal_RelOp(ASSERT_VALARG_TP assertions, Gen
             noway_assert(vnStore->ConstantValue<size_t>(vnCns) == 0);
             op1->AsIntCon()->gtIconVal = 0;
         }
+        else if (op1->TypeGet() == TYP_BYREF)
+        {
+            op1->ChangeOperConst(GT_CNS_INT);
+            op1->AsIntCon()->gtIconVal = vnStore->ConstantValue<target_size_t>(vnCns);
```

Comment:
I believe this should be using `target_ssize_t`. Our contract for `gtIconVal` for `TYP_INT` (which `TYP_BYREF` size-wise is when cross compiling to 32 bit on 64 bit) is that the value must be sign-extended. Assigning from an unsigned type will zero-extend it, which is incorrect.

Reply:
Fine I will chnage it to target_ssize_t.  It shouldn't really matter, since the only byref constant that I think we will see is 0.

Reply:
If you are saying it's only 0 should we add an assert here? Or noway_assert?

Reply:
I think we can also see byref consts with RVA statics these days, for example the following code:
```cs
public static ReadOnlySpan<byte> RVAStatic => new byte[] { 1, 2, 3 };
private static ref readonly byte ReturnByrefConst() => ref MemoryMarshal.GetReference(RVAStatic);
```
Gets us the following tree (among many copies...):
```scala
N003 (  6, 13) [000039] -A------R---              *  ASG       byref  $100
N002 (  3,  2) [000038] D------N----              +--*  LCL_VAR   byref  V09 tmp9         d:1 $100
N001 (  2, 10) [000058] H-----------              \--*  CNS_INT(h) byref  0xd1ffab1e static Fseq[039058C6F2C0CB492C533B0A4D14EF77CC0F78ABCCCED5287D84A1A2011CFB81] $100
```
Actually we propagate this constant in assertion prop as `long` (64 bit) for reasons I do not quite understand:
```scala
After constant propagation on [000080]:
STMT00001 (IL   ???...  ???)
N005 (  5,  4) [000085] -A----------              *  COMMA     void   $c2
N003 (  5,  4) [000081] -A------R---              +--*  ASG       byref  $100
N002 (  3,  2) [000079] D------N----              |  +--*  LCL_VAR   byref  V05 tmp5         d:1 $100
               [000093] H-----------              |  \--*  CNS_INT(h) long   0xd1ffab1e static $100
N004 (  0,  0) [000084] ------------              \--*  NOP       void   $c2
```
But we can end up with constant byrefs.

Reply:
I wouldn't add an assert that it must be zero.  
Perhaps we can see this with a RVA static as well as @SingleAccretion pointed out.

Reply:
Using a target_ssize_t causes an assert to be hit in valuenum.h, line 764.



`assert(&typeid(T) == &typeid(size_t)); // We represent ref/byref constants as size_t's.`

Reply:
Ehh, looks like it will need to use a `static_cast<target_ssize_t>` then :(.

Reply:
We are currently using size_t to store both refs and byrefs constants values in VN.
I don't believe that cross compiling is impacted here either.

Reply:
> We are currently using size_t to store both refs and byrefs constants values in VN.

Yes, I understand. But here we are converting between the way VN stores them and the IR does, and the IR always stores them as signed.

My concern comes from the following scenario: say the original IR had a byref constant in it, and we have a 64 bit host and a 32 bit target, which means that we sign-extended this constant (say the upper bits are all ones). Now VN saves this constant in the map (doesn't actually matter how it does it, what matter is how we retrieve it) and then the code rehydrates it into an IR constant again, but now zero-extending - we end up with a different value, i. e. the result of `IconValue() < 0` will be different. And this all of course only matter for the cross-compilation scenario because for native compilation there are no extensions, values are stored according to their `TYP` widths.

Reply:
I don't think that code can rely upon the order of  constant byref pointers, only equality and inequality should be used when comparing byref pointers.  But I have added the extra compares in order to get this checked in.

I note that there seems to some other nonsense going on when we are cross targeting from 64-bit to 32-bit.
I don't know if it impacts anything or if it is just some dead code.

Take a look at the method  `TruncateOrSignExtend32()` in gentree.h line 3173
It is checking a GT_CNS_INT node for the presence of the flag GTF_UNSIGNED
This flag is not permitted on a GT_CNS_INT node and will always return false.

Reply:
> I don't think that code can rely upon the order of constant byref pointers, only equality and inequality should be used when comparing byref pointers. But I have added the extra compares in order to get this checked in.

Yes I agree that we ought not to be checking byrefs for their sign, but note that in my example equality is also impacted, i. e. `originalNode->IconValue() == rehydratedNode->IconValue()` will return `false`.

> I note that there seems to some other nonsense going on when we are cross targeting from 64-bit to 32-bit.
I don't know if it impacts anything or if it is just some dead code.

> Take a look at the method TruncateOrSignExtend32() in gentree.h line 3173
It is checking a GT_CNS_INT node for the presence of the flag GTF_UNSIGNED
This flag is not permitted on a GT_CNS_INT node and will always return false.

I've been meaning to delete that method for a little while now :). My understanding of it aligns with yours in that the `GTF_UNSIGNED` check is dead. The only way, in theory (modulo overlapping flags), to hit is if something bashes an oper to a const without resetting the flags, which in itself is incorrect.